public/app/plugins/panel/canvas/editor/cad/utils.ts

```diff
@@ -78,22 +123,103 @@ function addTextElement(entity: ITextEntity, entityLayer: ILayer, canvasLayer: F
       bottom: entity.startPoint.y * TEMP_MULTIPLIER,
       left: entity.startPoint.x * TEMP_MULTIPLIER,
       rotation: -entity.rotation,
-      height: entity.textHeight * TEMP_MULTIPLIER * 2.5,
+      height: entity.textHeight * TEMP_MULTIPLIER,
       width: entity.textHeight * entity.text.length * TEMP_MULTIPLIER,
     },
     config: {
       text: { fixed: entity.text },
-      size: entity.textHeight * TEMP_MULTIPLIER,
+      size: (entity.textHeight * TEMP_MULTIPLIER) / 2,
       color: {
-        fixed: fromColorRepr(entity.color, entityLayer),
+        fixed: hexFromColorRepr(entity.color, entityLayer),
       },
     },
   };
 
-  addElement(newTextItem, newElementOptions, canvasLayer);
+  canvasLayer.addElement(new ElementState(newTextItem, newElementOptions, canvasLayer));
+}
+
+function lineFromVertices(vertices: Array<{ x: number; y: number }>): Line {
+  if (vertices.length !== 2) {
+    throw new Error(`unexpected number of vertices: expected 2, got ${vertices.length}`);
+  }
+
+  const dy = vertices[1].y - vertices[0].y;
+  const dx = vertices[1].x - vertices[0].x;
+
+  const theta = (Math.atan2(dy, dx) * 180) / Math.PI;
+  const length = Math.hypot(dx, dy);
+
+  return {
+    theta,
+    length,
+    midPoint: {
+      x: (vertices[0].x + vertices[1].x) / 2,
+      y: (vertices[0].y + vertices[1].y) / 2,
+    },
+  };
 }
 
-function fromColorRepr(color: number | undefined, cadLayer?: ILayer): string {
+function addLineElement(entity: ILineEntity, entityLayer: ILayer, canvasLayer: FrameState) {
+  const line = lineFromVertices(entity.vertices);
+  const lineWeight = entity.lineweight !== undefined && entity.lineweight !== 0 ? entity.lineweight : 0.25;
+
+  const newLineItem: CanvasElementItem<LineConfig, LineData> = canvasElementRegistry.get('line');
+  let newElementOptions: CanvasElementOptions = {
+    ...newLineItem.getNewOptions(),
+    type: newLineItem.id,
+    name: '',
+    constraint: BOTTOM_LEFT_CONSTRAINT,
+    placement: {
+      bottom: (line.midPoint.y - lineWeight / 2) * TEMP_MULTIPLIER,
+      left: (line.midPoint.x - line.length / 2) * TEMP_MULTIPLIER,
+      height: lineWeight,
```

Comment:
Should this also have the scale multiplier?

Reply:
It shouldn't, we're actually scaling everything to better match CAD's render based on our minimum line width.